Add framework for singleton instructions.

jonmeow opened the pull request:
Adds a singleton framework, and converts `File` and `InstId::Print` to demonstrate functionality. Moves various functions from `ids.h` to `ids.cpp` because `Inst::Print` needs the file if `singleton_insts.h` is split out, so the small bit of cleanup feels consistent.

I added `Inst::MakeSingleton` because getting the type of the instruction to make from an `InstKind` felt too hard. Singleton instructions follow a basic structure, so I'm just putting that instruction structure into `Inst`. Previously we required macros to do this, and I'm trying to remove macro dependencies.

I'm trying to remove builtin/singleton-related functionality from `InstId` in order to get a clearer boundary for the functionality. The other builtin functions should be removed as part of the bigger migration, but I'm trying to carefully scope changes to verify agreement on the singleton approach in use first.

This provides `InstT::SingletonInstId` because that'll often be written as `SemIR::TypeType::SingletonInstId`. The alternative of something like `SemIR::SingletonInstId<SemIR::TypeType>` is just a little more verbose due to the repeated `SemIR`, and it's more consistent with `TypeType::Kind`.

An alternative I considered was consolidating singleton information to `InstKind`. This felt challenging because of the `InstId::BuiltinTypeType` and similar values. Maintaining those would turn into something like `InstKind::IsSingleton()` and `InstId::Singleton<TypeType>`, which didn't feel like as good a split. `InstId::Print` remains aware of singletons, but I'm hoping to remove other builtin-related calls from `InstId`.

Another thing I considered was adding `.is_singleton = true` to `InstKind::Definition`. I don't think we could rely on that to get `InstId::Singleton<TypeType>` set up as `constexpr`, though. At that point, I think it'd mainly be _just_ a comment-like annotation, maybe validated with `CHECK` but not having any effect on its own. So I decided not to do that, just adding comments instead.

Sidenotes:

- "singleton" naming was discussed [on #toolchain](https://discord.com/channels/655572317891461132/655578254970716160/1308870233729269781).
- The TODO in file.h about possibly excluding other things than singletons seems moot. That's for raw IR, and we're much more focused on textual IR these days.

jonmeow committed:
Add framework for singleton instructions.

geoffromer commented on toolchain/sem_ir/singleton_insts.h:
```diff
@@ -0,0 +1,66 @@
+    InstKind::WitnessType,
+};
+
+namespace Internal {
```
As an alternative to consider, would it work to declare _but not define_ the public API, then have an `// Implementation details only below here` comment, and put `GetSingletonInstIndex` below that, followed by the definitions of the public API? That way we don't have to stick this implementation detail in the middle of the API, and instead the two halves of the file read like a conventional .h/.cc pair.

jonmeow replied:
Can you give an example code snippet? How does that work with `requires` and `constexpr`?

geoffromer replied:
Something like this:

```c++
// Returns true if the InstKind is a singleton.
static constexpr auto IsSingletonInstKind(InstKind kind) -> bool;

// Provides the InstId for singleton instructions. For example,
// `InstId error_inst_id = SingletonInstId<ErrorInst>;`.
template <typename InstT>
  requires(IsSingletonInstKind(InstT::Kind))
static constexpr InstId SingletonInstId;


// Implementation details only below here

// Returns the index for a singleton instruction, or -1 if it's not a singleton.
static constexpr auto GetSingletonInstIndex(InstKind kind) -> int32_t {
  for (int32_t i = 0; i < static_cast<int32_t>(SingletonInstKinds.size());
       ++i) {
    if (SingletonInstKinds[i] == kind) {
      return i;
    }
  }
  return -1;
}

static constexpr auto IsSingletonInstKind(InstKind kind) -> bool {
  return Internal::GetSingletonInstIndex(kind) >= 0;
}

template <typename InstT>
  requires(IsSingletonInstKind(InstT::Kind))
static constexpr InstId SingletonInstId =
    InstId(Internal::GetSingletonInstIndex(InstT::Kind));
```

Some of the details may need to be tweaked (I think maybe redeclarations are supposed to omit `static`?) but I believe something along those lines will work.

jonmeow replied:
Done. Also adjusted to `inline` instead of `static` for functions.

geoffromer commented on toolchain/sem_ir/inst.h:
```diff
@@ -127,6 +128,18 @@ concept InstLikeType = requires { sizeof(InstLikeTypeInfo<T>); };
 //   data where the instruction's kind is not known.
 class Inst : public Printable<Inst> {
  public:
+  // Makes the an instruction for a singleton. This exists to support simple
```
```suggestion
  // Makes an instruction for a singleton. This exists to support simple
```

geoffromer commented on toolchain/sem_ir/typed_insts.h:
```diff
@@ -47,7 +47,7 @@ namespace Carbon::SemIR {
 // Used for the type of patterns that do not match a fixed type.
 //
-// TODO: Annotate as a builtin.
+// This is a singleton instruction.
```
Would it be possible to add something like
```
static constexpr InstId SingletonId = SingletonInstId<AutoType>;
```
so that we can say `SemIR::AutoType::SingletonId` instead of `SemIR::SingletonInstInd<SemIR::AutoType>`?

jonmeow replied:
Done, with some refactoring to make it the favored path.

dwblaikie committed:
element index init with vptr (#4565)

Not sure if this is the most robust way to do it - I guess the
alternative is doing name lookup into the dest struct fields too?

danakj committed:
Update links to the DiagnosticConsumers (#4580)

The ConsoleDiagnosticConsumer and ErrorTrackingDiagnosticConsumer have
moved to new places.

danakj committed:
Refresh the ValueStore api description in idioms.md (#4586)

There is no `Set` method, so remove that, and include `AddDefaultValue`
instead.

danakj committed:
Update idioms.md with changes to match the code now (#4587)

- Rename TypedInstArgsInfo references to InstLikeTypeInfo. The type was
renamed in 07efa026de540.
- Update and correct the link to search for ValueStore (and similar) in
the Context class. We must avoid the google-doc-style checks rewriting
`repo:` to `repository:` in the URL.
- Mention the existance of many types of Store collections now.
- Remove pre-C++20 idioms in Field detection. Correct the concept based
idioms to work.
- Fix code indenting consistency.

---------

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

josh11b committed:
Mark instructions that can be deduced through in `typed_insts.h` (#4588)

Co-authored-by: Josh L <josh11b@users.noreply.github.com>

josh11b committed:
Make facet type deduction more restrictive and correct (#4589)

Previously it would allow interface mismatches. We only need to support
the case where there is a single interface, though, which makes checking
much more straightforward.

Co-authored-by: Josh L <josh11b@users.noreply.github.com>

josh11b committed:
Facet member access (#4371)

Adds `FacetAccessWitness` instruction and uses it in `member_access.cpp`
to support accessing members of facets. Still to do: interface witness
access is producing runtime values when it should produce symbolic
values.

---------

Co-authored-by: Josh L <josh11b@users.noreply.github.com>

zygoloid committed:
Switch most of `ImportRefResolver` to non-member functions (#4584)

In preparation for further refactoring, switch away from member
functions for most of `ImportRefResolver`.

Split `ImportRefResolver` into a context class that exposes the value
stores for the source and destination files, and a derived class that
maintains a worklist. The idea is to statically enforce that functions
that take `ImportContext` cannot accidentally add new work, because they
don't have access to the work queue.

josh11b committed:
Simplify logic in `StringifyTypeExpr` using `push_string` (#4561)

Follow-on to #4511.

---------

Co-authored-by: Josh L <josh11b@users.noreply.github.com>

josh11b committed:
Link to video for 2024 CppNorth talk (#4578)

* Change format:
  * previously there was a single link per talk, the video if available
  * now there are separate "video" and "slide" links
* Adds link to video for 2024 CppNorth talk
* Adds link to slides for 2024 LLVM Developers' Meeting
* Restores links to slides from past years

---------

Co-authored-by: Josh L <josh11b@users.noreply.github.com>
Co-authored-by: Jon Ross-Perkins <jperkins@google.com>
Co-authored-by: Chandler Carruth <chandlerc@gmail.com>

chandlerc committed:
Improve hashtable prefetching (#4585)

I had removed most but not all of the hashtable prefetching during
development because I wasn't confident in the benchmarking results.
However, I never revisited this once the benchmarking infrastructure
improved and there were solid and stable results.

This factors the two interesting prefetch patterns I've seen for this
style of hashtable into helpers that are always called, and provides
macros that can be used during the build to configure exactly which
prefetch strategies are enabled.

Benchmarking these and gaining confidence is very frustrating -- even
now with the improved infrastructure, the noise is much higher than I
would like. But it seems clear that *some* prefetching is a significant
win. It also seems like enabling both results in too much prefetch
traffic. And the entry group prefetch appears to be significantly more
effective, both for the most interesting of the microbenchmarks and
maybe most importantly for our compilation benchmarks. There, AMD is
helped substantially and M1 seems to be helped some (although harder to
measure).

AMD server benchmark numbers:
```
name                                              old cpu/op   new cpu/op   delta
BM_CompileAPIFileDenseDecls<Phase::Lex>/256       35.0µs ± 2%  34.2µs ± 2%  -2.40%  (p=0.000 n=20+19)
BM_CompileAPIFileDenseDecls<Phase::Lex>/1024       156µs ± 2%   151µs ± 2%  -3.18%  (p=0.000 n=20+19)
BM_CompileAPIFileDenseDecls<Phase::Lex>/4096       625µs ± 1%   605µs ± 1%  -3.22%  (p=0.000 n=19+18)
BM_CompileAPIFileDenseDecls<Phase::Lex>/16384     2.79ms ± 1%  2.69ms ± 2%  -3.67%  (p=0.000 n=17+19)
BM_CompileAPIFileDenseDecls<Phase::Lex>/65536     12.1ms ± 1%  11.6ms ± 1%  -4.30%  (p=0.000 n=17+18)
BM_CompileAPIFileDenseDecls<Phase::Lex>/262144    56.6ms ± 1%  53.8ms ± 1%  -5.00%  (p=0.000 n=18+17)
BM_CompileAPIFileDenseDecls<Phase::Parse>/256     61.1µs ± 2%  61.7µs ± 1%  +0.87%  (p=0.000 n=19+19)
BM_CompileAPIFileDenseDecls<Phase::Parse>/1024     288µs ± 1%   290µs ± 1%  +0.55%  (p=0.004 n=20+20)
BM_CompileAPIFileDenseDecls<Phase::Parse>/4096    1.16ms ± 1%  1.16ms ± 1%  -0.54%  (p=0.000 n=17+19)
BM_CompileAPIFileDenseDecls<Phase::Parse>/16384   4.98ms ± 1%  4.91ms ± 1%  -1.39%  (p=0.000 n=20+19)
BM_CompileAPIFileDenseDecls<Phase::Parse>/65536   20.9ms ± 1%  20.5ms ± 1%  -1.86%  (p=0.000 n=20+19)
BM_CompileAPIFileDenseDecls<Phase::Parse>/262144  92.1ms ± 1%  90.2ms ± 1%  -2.12%  (p=0.000 n=18+20)
BM_CompileAPIFileDenseDecls<Phase::Check>/256     1.16ms ± 2%  1.16ms ± 1%    ~     (p=0.931 n=19+19)
BM_CompileAPIFileDenseDecls<Phase::Check>/1024    2.17ms ± 2%  2.16ms ± 1%    ~     (p=0.247 n=20+19)
BM_CompileAPIFileDenseDecls<Phase::Check>/4096    6.07ms ± 1%  6.04ms ± 1%  -0.48%  (p=0.007 n=19+19)
BM_CompileAPIFileDenseDecls<Phase::Check>/16384   22.4ms ± 1%  22.2ms ± 1%  -0.99%  (p=0.000 n=20+19)
BM_CompileAPIFileDenseDecls<Phase::Check>/65536   93.3ms ± 1%  92.2ms ± 1%  -1.23%  (p=0.000 n=20+18)
BM_CompileAPIFileDenseDecls<Phase::Check>/262144   400ms ± 1%   391ms ± 1%  -2.15%  (p=0.000 n=20+18)
```

danakj committed:
Add "Whether to" in the description of 'debug-info' switch (#4592)

The switch defaults to true, so it's displayed as `--no-debug-info` in
the help. The description should be agnostic about whether the flag will
enable or disable the behaviour.

dwblaikie committed:
lazy field index (#4514)

We considered a couple of other options for this:
* https://github.com/carbon-language/carbon-lang/pull/4515 Keep the
`ElementIndex` numbering vptr-ignorant, and do +1 offsets as needed -
seems subtle/easy to miss
* https://github.com/carbon-language/carbon-lang/pull/4517 Always have a
zeroth element in the object representation, make it zero-size in the
case of no-vptr - @zygoloid was concerned this would add overhead
especially to stateless objects used in type-trait-like things.

But currently moving forward with this direction - of initializing field
indexes with an invalid value until the end of the class definition,
then assigning field indexes during construction of the class's object
representation struct type. This direction might reinforce/help avoid
premature access to the object representation before the class is
complete, and give a single place where class layout is done (at class
completion) if we want to add more options there, such as class layout
optimizations, etc.

This patch still has problems with object initialization (that #4515
does not have/does address) but does address normal `obj.member` access
correctly.

---------

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

jonmeow committed:
Refactor command line errors to mirror diagnostic style (#4568)

This changes to an `Error` return to let the driver do the "error: "
prefix, except for one case with `help` that needs more work to change
(I'm not planning on picking up that TODO). It also changes
capitalization, backtick use, and a few minor punctuation things to try
to better match the diagnostic style.

This also adds `Error` matchers so that the changes to command line
testing are clearer.

josh11b committed:
Symbolic aggregate access (#4590)

Co-authored-by: Josh L <josh11b@users.noreply.github.com>
Co-authored-by: Richard Smith <richard@metafoo.co.uk>

zygoloid committed:
Add documentation for the three most common kinds of instruction operand (#4594)

Co-authored-by: Carbon Infra Bot <carbon-external-infra@google.com>
Co-authored-by: Dana Jansens <danakj@orodu.net>

zygoloid committed:
Don't print a comma in a two-item list. (#4596)

dwblaikie committed:
disallow impl without base (#4583)

josh11b committed:
Refactor `StringifyTypeExpr` (#4597)

* Make the step stack into a class
* Make the operations on the stack (pushing, popping, test for done)
into methods on the stack class.
* Add more kinds of steps (array bound and name).
* Rewrite cases to use the new kinds of steps. Afterward, none use
`Step::Next()` or the step index, so those get removed.
* Add another convenience method `PushTypeId`.
* Remove the `SemIR::File&` member from the steps, since it doesn't
change.

Hopefully using `step_stack.Push`... calls makes it clear that they are
resolved in the reverse order they are executed.

---------

Co-authored-by: Josh L <josh11b@users.noreply.github.com>

danakj committed:
Use a switch in StringifyTypeExpr (#4598)

Get a compiler error when a case isn't handled instead of a runtime
CHECK failure.

bricknerb committed:
Add a test that verifies we prefer unqualified name lookup from a a non lexical scope over a lexical scope (#4591)

This adds missing test coverage.
See discussion in
https://github.com/carbon-language/carbon-lang/pull/4574.

zygoloid committed:
Track the type as written in `BaseDecl` and `AdaptDecl`. (#4564)

Represent the type as an `InstId` rather than as a `TypeId` to preserve
how it was written and better support tracking its value in a generic.
Add accessors to `Class` to get the base and adapted type to reduce code
duplication, and add `TypeStore::GetObjectRepr` to make it easier to map
from a type to its possibly-adapted object representation type. In
passing, also move `GetIntTypeInfo` and `GetUnqualifiedType` into
`TypeStore`.

This fixes specifics of generic adapters to properly look at the
specific adapted type, and also fixes importing of adapters.

---------

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

zygoloid committed:
Include the complete type witness for a class in its SemIR output. (#4599)

This seems slightly redundant for a locally-defined class, where there
will be a complete_type_witness instruction earlier in the class, but is
important for imported classes, where we're currently doing the wrong
thing in a way that's invisible in formatted SemIR.

zygoloid committed:
Fix importing of the complete_type_witness for a generic class. (#4600)

We need to create an instruction on import to attach the generic
constant value to.

bricknerb committed:
Add a basic test for name poisoning (#4572)

Demonstrate where failures should happen.

---------

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>
Co-authored-by: David Blaikie <dblaikie@gmail.com>
Co-authored-by: Dana Jansens <danakj@orodu.net>
Co-authored-by: Richard Smith <richard@metafoo.co.uk>
Co-authored-by: josh11b <15258583+josh11b@users.noreply.github.com>
Co-authored-by: Josh L <josh11b@users.noreply.github.com>
Co-authored-by: Chandler Carruth <chandlerc@gmail.com>
Co-authored-by: Carbon Infra Bot <carbon-external-infra@google.com>
Co-authored-by: Geoff Romer <gromer@google.com>

zygoloid committed:
Fix impl lookup to look in the arguments of a specific. (#4601)

zygoloid committed:
Include the call arguments in the location of a call. (#4602)

Underline the entire call in diagnostics, not only the portion up to the
`(`.

bricknerb committed:
[NFC] In DeclNameStack::AddName(), use NameScope::AddRequired() instead of duplicating its logic (#4610)

This is a pure code deduplication change.

danakj committed:
Expose InstKind::FromInt for Inst instead of InstKind::Make (#4611)

The Inst type will type erase a specific typed instruction by storing
the kind as an integer. It does this by calling InstKind::AsInt on a
runtime or compile-time InstKind. Then it returns the kind as InstKind
by reconstituting it from the integer.

Currently it does a cast to a raw enumerator and then calls
InstKind::Make. However Make is designed to be more of an internal
detail. The more clearly paired inverse operation is InstKind::FromInt,
which is documented as being intended to be exposed by derived classes
like InstKind.

danakj committed:
CHECK that PreCheck and GetCheckUnit are not called twice (#4608)

If they were called twice for a CompilationUnit, they would destroy
objects that they created and returned a pointer to, leaving a dangling
pointer somewhere else.

jonmeow committed:
Update toolchain/sem_ir/inst.h

Co-authored-by: Geoff Romer <gromer@google.com>

jonmeow committed:
SingletonInstId

jonmeow committed:
Switch from bazeliskrc to bazelversion (#4612)

The only documentation I can find for `.bazelversion` is in
https://github.com/bazelbuild/bazelisk/blob/master/README.md. However,
`bazel` will error out if it doesn't match the `.bazelversion`. For
example:

```
╚╡/usr/bin/bazel build :all
ERROR: The project you're trying to build requires Bazel 7.3.0 (specified in [elided]/carbon-lang/.bazelversion), but it wasn't found in /usr/bin.
```

Switching to this because the `MODULE.bazel` requires tighter version
pinning in order to avoid churn, and this should help catch mistakes
early.

In `MODULE.bazel`, demote mention of the version issue because it should
only really occur now when the version is being deliberately changed, so
the connection should be more apparent.

zygoloid committed:
Support for stringifying int values used as generic arguments. (#4614)

zygoloid committed:
Factor out machinery for forming int type literals. (#4616)

Use it in the remaining few places where we currently hardcode `i32`: as
the index type in array indexing, as the type for literals in `if`
expressions, and as a valid return type for `Run`.

In preparation for changing `Core.Int` to be a class.

---------

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

jonmeow committed:
decl

danakj committed:
Use a Timings* in place of optional<Timings>* (#4607)

The current representation has two nested presence indicatators (the
optional bool, the null pointer). Currently the pointer is never null,
but the style guide suggests that T* should be used for parameters that
may or may not be present, so we do not need the optional here.

> When passing an object's address as an argument, use a reference
> unless one of the following cases applies:
>
> - If the parameter is optional, use a pointer and document that it
>   may be null.

Once the parameter is just a pointer, the ScopedTiming field does not
need an optional either, and can just store the pointer.

It would be more preferable to have an optional representation of a
sometimes-null pointer like optional<T&> to describe a sometimes-null
pointer, as this would allow clearer runtime diagnostics when used
incorrectly (a check failure in unwrapping) and would be better
self-documenting through syntax instead of a comment. But we do not
currently have such a primitive.

jonmeow committed:
Re-add Make
